File: ui/src/stores/animationStore/playback.ts
/**
 * Playback Actions
 *
 * Core playback operations: play, pause, toggle, frame stepping.
 */

import { clearMaskPathCacheOnSeek } from "@/services/effects/maskRenderer";
import { clearTimeEffectStateOnSeek } from "@/services/effects/timeRenderer";
import { usePlaybackStore } from "../playbackStore";
import type { AnimationStoreAccess } from "./types";

/**
 * Start playback
 */
export function play(store: AnimationStoreAccess): void {
  const playback = usePlaybackStore();
  if (playback.isPlaying) return;

  const comp = store.getActiveComp();
  if (!comp) return;

  playback.play(
    comp.settings.fps,
    comp.settings.frameCount,
    comp.currentFrame,
    (frame: number) => {
      comp.currentFrame = frame;
    },
  );

  // Note: playbackStore.isPlaying is already set to true by playback.play()
}

/**
 * Pause playback
 */
export function pause(store: AnimationStoreAccess): void {
  const playback = usePlaybackStore();
  playback.stop();
  // Note: playbackStore.isPlaying is already set to false by playback.stop()
}

/**
 * Toggle playback state
 */
export function togglePlayback(store: AnimationStoreAccess): void {
  const playback = usePlaybackStore();
  if (playback.isPlaying) {
    pause(store);
  } else {
    play(store);
  }
}

/**
 * Set current frame
 * Clears temporal effect state when frame changes non-sequentially.
 */
export function setFrame(store: AnimationStoreAccess, frame: number): void {
  const comp = store.getActiveComp();
  if (!comp) return;

  // Validate frame (NaN bypasses Math.max/min)
  if (!Number.isFinite(frame)) return;

  const newFrame = Math.max(0, Math.min(frame, comp.settings.frameCount - 1));

  // Clear temporal state if frame changes by more than 1 (non-sequential)
  const frameDelta = Math.abs(newFrame - comp.currentFrame);
  if (frameDelta > 1) {
    clearTimeEffectStateOnSeek();
    clearMaskPathCacheOnSeek();
  }

  comp.currentFrame = newFrame;
}

/**
 * Advance to next frame
 */
export function nextFrame(store: AnimationStoreAccess): void {
  const comp = store.getActiveComp();
  if (!comp) return;
  if (comp.currentFrame < comp.settings.frameCount - 1) {
    comp.currentFrame++;
  }
}

/**
 * Go to previous frame
 */
export function prevFrame(store: AnimationStoreAccess): void {
  const comp = store.getActiveComp();
  if (!comp) return;
  if (comp.currentFrame > 0) {
    comp.currentFrame--;
  }
}

/**
 * Jump forward or backward by N frames
 */
export function jumpFrames(store: AnimationStoreAccess, n: number): void {
  const comp = store.getActiveComp();
  if (!comp) return;

  // Validate n (NaN bypasses Math.max/min)
  if (!Number.isFinite(n)) return;

  const newFrame = Math.max(
    0,
    Math.min(comp.currentFrame + n, comp.settings.frameCount - 1),
  );
  if (Math.abs(n) > 1) {
    clearTimeEffectStateOnSeek();
    clearMaskPathCacheOnSeek();
  }
  comp.currentFrame = newFrame;
}

File: src/lattice/State/Animation/Playback.hs
-- |
-- Module      : Lattice.State.Animation.Playback
-- Description : Playback operations for animation store
-- 
-- Migrated from ui/src/stores/animationStore/playback.ts
-- Pure functions for frame operations - playback control (play/pause) handled at FFI boundary
--

{-# LANGUAGE OverloadedStrings #-}

module Lattice.State.Animation.Playback
  ( setFrame
  , nextFrame
  , prevFrame
  , jumpFrames
  , shouldClearTemporalState
  ) where

import Data.Maybe (Maybe)
import Data.Text (Text)
import qualified Data.Text as T
import Lattice.Types.Primitives (validateFinite, validateNonNegative)
import Lattice.Types.Project (Composition(..), CompositionSettings(..))

-- ============================================================================
-- FRAME OPERATIONS
-- ============================================================================

-- | Set current frame in composition
-- Pure function: takes composition and frame, returns new composition with updated currentFrame
-- Clears temporal effect state when frame changes non-sequentially (indicated by return value)
setFrame ::
  Composition -> -- Current composition
  Double -> -- Target frame number
  Either Text (Composition, Bool) -- (Updated composition, should clear temporal state), or error
setFrame comp frame =
  if not (validateFinite frame)
    then Left ("setFrame: Invalid frame value (must be finite): " <> T.pack (show frame))
    else
      let
        settings = compositionSettings comp
        frameCount = compositionSettingsFrameCount settings
        currentFrame = compositionCurrentFrame comp
        
        -- Clamp frame to valid range [0, frameCount - 1]
        newFrame = max 0.0 (min frame (frameCount - 1))
        
        -- Check if frame change is non-sequential (delta > 1)
        frameDelta = abs (newFrame - currentFrame)
        shouldClear = frameDelta > 1.0
        
        -- Update composition
        updatedComp = comp {compositionCurrentFrame = newFrame}
      in
        Right (updatedComp, shouldClear)

-- | Advance to next frame
-- Pure function: takes composition, returns new composition with incremented currentFrame
nextFrame ::
  Composition -> -- Current composition
  Either Text Composition -- Updated composition, or error
nextFrame comp =
  let
    settings = compositionSettings comp
    frameCount = compositionSettingsFrameCount settings
    currentFrame = compositionCurrentFrame comp
  in
    if currentFrame < frameCount - 1
      then Right (comp {compositionCurrentFrame = currentFrame + 1})
      else Right comp  -- Already at last frame, no change

-- | Go to previous frame
-- Pure function: takes composition, returns new composition with decremented currentFrame
prevFrame ::
  Composition -> -- Current composition
  Either Text Composition -- Updated composition, or error
prevFrame comp =
  let
    currentFrame = compositionCurrentFrame comp
  in
    if currentFrame > 0
      then Right (comp {compositionCurrentFrame = currentFrame - 1})
      else Right comp  -- Already at first frame, no change

-- | Jump forward or backward by N frames
-- Pure function: takes composition and frame delta, returns new composition with updated currentFrame
-- Clears temporal effect state when |n| > 1 (indicated by return value)
jumpFrames ::
  Composition -> -- Current composition
  Double -> -- Frame delta (positive = forward, negative = backward)
  Either Text (Composition, Bool) -- (Updated composition, should clear temporal state), or error
jumpFrames comp n =
  if not (validateFinite n)
    then Left ("jumpFrames: Invalid frame delta (must be finite): " <> T.pack (show n))
    else
      let
        settings = compositionSettings comp
        frameCount = compositionSettingsFrameCount settings
        currentFrame = compositionCurrentFrame comp
        
        -- Calculate new frame and clamp to valid range [0, frameCount - 1]
        newFrame = max 0.0 (min (currentFrame + n) (frameCount - 1))
        
        -- Check if jump is non-sequential (|n| > 1)
        shouldClear = abs n > 1.0
        
        -- Update composition
        updatedComp = comp {compositionCurrentFrame = newFrame}
      in
        Right (updatedComp, shouldClear)

-- ============================================================================
-- HELPER FUNCTIONS
-- ============================================================================

-- | Check if temporal state should be cleared based on frame delta
-- Pure function: takes old frame and new frame, returns True if delta > 1
shouldClearTemporalState ::
  Double -> -- Old frame
  Double -> -- New frame
  Bool -- True if temporal state should be cleared
shouldClearTemporalState oldFrame newFrame =
  abs (newFrame - oldFrame) > 1.0
